File: SDP_Assignments/Uncertain_T/Uncertain_python/Uncertainty/UncertainTests/GaussianTests.py
import numpy as np 
import random
import sys
import os
currentdir = os.path.dirname(os.path.realpath(__file__))
parentdir = os.path.dirname(currentdir)
sys.path.append(parentdir)
from Uncertain.Gaussian import Gaussian
from Uncertain.Sampler import Sampler
from Uncertain.Extensions import Extensions 
from Uncertain.uncertain import Operator 

class GaussianTests():
    def gaussian_sample(self):
        x = Gaussian(5.0,2.0)
        x=x.get_support()
        sampler=Sampler()
        sampler.create(x)
        for i in sampler.take(100):
            assert(i >= -3 and i <= 13)

    def gaussian_mean(self):
        """[summary]
        """
        sum_value = 0.0
        x = Gaussian(5.0,1.0)
        x=x.get_support()
        sampler=Sampler()
        sampler.create(x)
        for i in sampler.take(100):
            sum_value += i
        sum_value /=100
        assert(sum_value>=4.6 and sum_value <=5.4)

    def gaussian_bnn_sample(self):
        x = Gaussian(1.0,1.0)
        y = Gaussian(4.0,1.0)
        z = Operator(x)+Operator(y)
        sampler=Sampler()
        sampler.create(z)
        for i in sampler.take(100):
            assert(i >=-3.0 and i<=13.0) 

    def gaussian_bnn_mean(self):
        x = Gaussian(1.0,1.0)
        y = Gaussian(4.0,1.0)
        z = Operator(x)+Operator(y)
        sum =0.0
        sampler=Sampler()
        sampler.create(z)
        for i in sampler.take(100):
            sum=sum+i
        sum = sum/100.0
        assert(sum>=4.11 and sum <=5.89)

    def gaussian_bernoulli_mean(self):
        x = Gaussian(1.0,1.0)
        y = Gaussian(3.0,2.0)
        z = Operator(x)>Operator(y)
        sampler=Sampler()
        sampler.create(z)
        k=0
        for i in sampler.take(100):
            if i<0:
                k=k+1
        assert (k>=13 and k<52)
    
    def gaussian_bernoulli_conditional(self):
        x = Gaussian(1.0,1.0)
        y = Gaussian(4.0,2.0)
        extns=Extensions()
        z1 = Operator(x)>Operator(y)
        z2= Operator(y)<Operator(x)
        z3 = Operator(y)>Operator(x)
        z4 = Operator(x)<Operator(y)
        if extns.pr(z1):
            print("X > Y evaluates true, incorrectly")
        if extns.pr(z2):
            print("Y < X evaluates true, incorrectly")
        if not extns.pr(z3):
            print("Y > X evaluates false, incorrectly")
        if not extns.pr(z4):
            print("X < Y evaluates false, incorrectly")

if __name__ == '__main__':
    test = GaussianTests()
    #test.gaussian_sample()
    #test.gaussian_mean()
    test.gaussian_bnn_sample()
    test.gaussian_bnn_mean()
    test.gaussian_bernoulli_mean()
    test.gaussian_bernoulli_conditional()

File: SDP_Assignments/Uncertain_T/cpp/GaussianTests.cpp
#include <iostream>
#include "uncertain.h"
#include "gaussian.h"
#include "ISampler.h"
// #include "random_primitive.h"

using namespace std;


// RandomPrimitive<T>::RandomPrimitive(){}

void gaussian_sample(){
    Gaussian x(5.0,2.0); 
    Sampler s;
    float* test = s.create(x.distribution,NUM_OF_SAMPLES); 
    for(int i=0; i<100; i++){
        cout << (test[i] >= -3 && test[i <= 13]) << endl; 
        if (!(test[i] >= -3 && test[i <= 13]) ){cout << "break" << endl; break;}
    }

}

void gaussian_mean(){

    float sum_value = 0.0; 
    Gaussian x(5,1); 
    Sampler s; 
    float* test = s.create(x.distribution,NUM_OF_SAMPLES); 
    for(int i=0; i<NUM_OF_SAMPLES; i++){
        sum_value += test[i]; 
    }

    sum_value = sum_value/NUM_OF_SAMPLES ; 
    cout << (sum_value >= 4.6 && sum_value <= 5.4) << endl; 
    if(!(sum_value >= 4.6 && sum_value <= 5.4)) {
        cout << "break" << endl; 
       
    }

}


void gaussian_bnn_sample(){


    // Gaussian z = x+y; 
    Uncertain<float> x = Gaussian(1,1);
    Uncertain<float> y = Gaussian(4,1);
    Uncertain<float> z = x+y;

  
    Sampler s; 
    float* test = s.create(z.distribution,NUM_OF_SAMPLES); 
    float* new_test = s.take(100); 

    for(int i=0; i<100; i++){
        cout << (new_test[i] >= -3.0 && new_test[i] <= 13) << endl; 

    }

}



void gaussian_bnn_mean(){


    // Gaussian z = x+y; 
    Uncertain<float> x = Gaussian(1,1);
    Uncertain<float> y = Gaussian(4,1);
    Uncertain<float> z = x+y;

    float sum = 0.0; 
    Sampler s; 
    float* test = s.create(z.distribution,NUM_OF_SAMPLES); 
    float* new_test = s.take(100); 

    for(int i=0; i<100; i++){
        sum += s.new_distribution[i]; 

    }
    sum = sum/100; 

    cout << (sum >= 4.11 && sum <= 5.89) << endl; 

}


void gaussian_bernoulli_mean(){

    Uncertain<float> x = Gaussian(1,1);
    Uncertain<float> y = Gaussian(4,2);
    Uncertain<float> z = x > y; 

    Sampler s; 
    float* test = s.create(z.distribution,NUM_OF_SAMPLES); 
    float* new_test = s.take(100); 

    int k = 0; 

    for(int i=0; i<100; i++){
        //workaround to get a bernoulli distribution 
        test[i] = abs(round(test[i])) ;
        if (test[i] > 1){ test[i] = 1;}

        if(test[i] >= 1){
            k += 1; 
        }
    }

    cout<< k << " " << (k>=13 && k < 52) << endl; 




}

void gaussian_bernoulli_conditional(){

    Uncertain<float> x = Gaussian(1,1);
    Uncertain<float> y = Gaussian(4,2);
    Uncertain<float> z = x < y;


    if((x>y).probability(1)){

        cout << "X > Y evaluates true, incorrectly" << endl;

    }
    if((x<y).probability(1)){

        cout << "X < Y evaluates true, incorrectly" << endl;

    }

    if(!(x>y).probability(0)){

        cout << "X > Y evaluates true, incorrectly" << endl;

    }

    if(!(x<y).probability(0)){

        cout << "X < Y evaluates true, incorrectly" << endl;

    }





}



int main(){

    Uncertain<float> x = Gaussian(11,0.03);
    Uncertain<float> y = Gaussian(10,0.04);


    gaussian_sample();
    gaussian_mean();
    gaussian_bnn_sample();
    gaussian_bnn_mean();
    gaussian_bernoulli_mean();
    gaussian_bernoulli_conditional();

    return 0;
}
